solr/core/src/java/org/apache/solr/search/DocSet.java

```diff
@@ -36,6 +36,21 @@
     assert this instanceof BitDocSet || this instanceof SortedIntDocSet;
   }
 
+  // we can't simply use a trivial static initializer "= new SortedIntDocSet" because it can lead to classloader deadlock
+  private static DocSet EMPTY;
+
+  /** An empty instance. */
+  public static DocSet empty() {
```

Comment:
If you do the singleton holder idiom then you don't need to have a null check (at the cost of an extra class, but the VM takes care of this quickly) [Effective Java 3rd Ed, Item 83] 

```
    private static class LazyHolder {
        static final DocSet INSTANCE = new SortedIntDocSet(new int[0]);
    }

    public static DocSet empty() {
        return LazyHolder.INSTANCE;
    }

Reply:
A fine suggestion. Do you think that approach is better?  For this case I find both no better/worse.

Reply:
I'll refer to the much wiser than myself, Josh Bloch (emphasis mine) -

> In summary, you should initialize most fields normally, not lazily. If you must initialize a field lazily in order to... break a harmful initialization circularity, then use the appropriate lazy initialization technique. For instance fields, it is the double-check idiom; **for static fields, the lazy initialization holder class idiom**. For instance fields that can tolerate repeated initialization, you may also consider the single-check idiom.

There's actually a lot more nuance in the surrounding paragraphs in the book, most of which I don't understand, but the summary appears to suggest that this is the fastest, simplest way.